defining codec specs

dericed opened the pull request:
Starting a process to better define how to define codec specs. This starts to detail how to create a Codec ID. And splits the concept of Description into Description, Citation, and Initialisation.

retokromer commented on codec_specs.md:
```diff
@@ -1,21 +1,59 @@
 Important Note:
 
-The intention behind this list is `NOT` to list all existing audio and video codecs, but rather to list those codecs that are `currently supported` in Matroska and therefore need a well defined codec ID so that all developers supporting Matroska will use the same ID. If you feel we missed support for a very important codec, please tell us on our development mailing list (matroska-devel at lists.m ...
+The intention behind this list is `NOT` to list all existing audio and video codecs, but rather to list those codecs that are `currently supported` in Matroska and therefore need a well defined codec ID so that all developers supporting Matroska will use the same ID. If you feel we missed support for a very important codec, please tell us on our development mailing list (cellar at ietf.org).
```
Why `NOT`? I don't think this in conforming RFC 2119.

mbunkus replied:
I agree with @retokromer. Let's correct it now: lower-case "not" without backticks. Additionally I'd remove the backticks around "currently supported". It's supposed to be an emphasis, I guess, but we usually use backticks for official terminology like element names.

retokromer commented on codec_specs.md:
```diff
@@ -1,21 +1,59 @@
+
+### Initialisation
+
+Each codec supported for storage in Matroska MUST have a defined Initialisation. If the codec does not require any form of Initialisation then `none` SHOULD be used to define the Initialisation. The Initialisation MUST document in detail how data MUST be stored within the CodecPrivate and CodecState Elements in order to initialize the codec.
```
«The Initialisation MUST document in detail how data MUST be stored» looks very weird to me. I suggest to replace the second MUST. – Yet this may be only my weird taste.

mbunkus replied:
@retokromer well, both MUSTs refer to different things: the first is a restriction on how to write the specs, the second a restriction on how multiplexers write the file. I'd rather keep both.

retokromer replied:
Exactly! I agree with you, @mbunkus, and for this very reason I would personally prefer to divide into two sentences these two behaviours.

mbunkus replied:
I'd be fine with that, of course.

dericed replied:
I removed the second MUST. I think that the Initialization statement for each codec itself should clarify the level of mandate.

mbunkus commented on codec_specs.md:
```diff
@@ -14,6 +14,46 @@ Important Note:
+Subtitle   | S_
+Button     | B_
+
+Each Codec ID MUST then use a `Major Codec ID` and may optionally use a `Minor Codec ID` and `Micro Codec ID` with a forward-slash (`/`) delimiter between each component.
```
Unfortunately this clashes a bit with existing practice from way back when Matroska was very young. We used the following monstrosities for AAC:

* `A_AAC/MPEG2/LC`
* `A_AAC/MPEG2/MAIN`
* `A_AAC/MPEG2/LC/SBR`
* `A_AAC/MPEG2/SSR`
* `A_AAC/MPEG4/LC`
* `A_AAC/MPEG4/LTP`
* `A_AAC/MPEG4/MAIN`
* `A_AAC/MPEG4/LC/SBR`
* `A_AAC/MPEG4/SSR`

As you can see some of them have one more layer. No `CodecPrivate` was used with them.

However, those IDs have been discontinued in favor of `A_AAC` with `CodecPrivate` 2005-12-07 (mkvmerge 1.6.5).

All other codec IDs that mkvmerge handles keep to major/minor/micro.

Looking forward I'm in favor of restricting official IDs to major/minor/micro.

dericed replied:
In some versioning systems, the fourth value is called `qualifier`. I suppose this term could fit here.

Also from codec_specs.md and https://matroska.org/technical/specs/codecid/index.html it doesn't seem possible to `know` that these Codec IDs are deprecated or what they are superseded by. Are these only discontinued by mkvmerge or by the Matroska format?

Should we have two other descriptors:
- Deprecation Date: optional but gives a date when the Matroska specification maintain deprecates a particular Codec ID.
- Superseded By: if Deprecation Date is used, then it should be highly recommended to say what Codec ID has superseded the deprecated one.

For players, they could then use the full registry for playback support, though writers should avoid using deprecated Codec IDs.

mbunkus replied:
Ugh. We've never deprecated the IDs officially, nor have we officially added the ID that's actually used — `A_AAC`. I clearly remember there being a discussion and consensus either on the mailing list or on the IRC channel about this change.

I don't know of a single Matroska muxer that still writes those old IDs.

I'd consider them de facto deprecated, with the specs page needing to catch up.

I like both of the descriptors you've proposed.

JeromeMartinez replied:
I think the proposal is too limitative, and we should let each ID free to define what is after "/".
I suggest:
"Codec ID Prefix"_"Major Codec ID"/ "Codec ID Specific"
(the "/" + right part being optional)

dericed replied:
I agree with Jerome's suggestion, I rewrote to use `Codec ID Prefix`, `Major Codec ID`, and `Codec ID Suffix`.

dericed commented:
Forced in a signification update. This describes a term `Codec Mapping` to describe the information required to define encoding support in Matroska. I updated the description of `Codec ID` and `Initialization` and added `Adoption Date`, `Deprecation Date`, and `Superseded By`.

Also what does `The private data is void` mean in the old specification document. For instance if using `A_MPEG/L3` then `the private data is void`. I'm confused by the use of the term `void` here. Does this mean that?
- a `CodecPrivate Element` MUST be present and be an `Empty Element`
- there MUST be no written `CodecPrivate Element`

Also I could use help describing when `CodecState Elements` should be used.

mbunkus commented:
Thanks. I'll do a re-review later or tomorrow.

> I'm confused by the use of the term void here.

Yeah :) It means the latter: the codec doesn't require private initialization data, and therefore a `CodecPrivate Element` MUST NOT exist. Technically an empty `CodecPrivate Element` would probably work just as well, however, there's no upside to allowing it apart from confusing players and trying to trigger bugs; therefore I'm in favor of disallowing that.

JeromeMartinez commented on codec_specs.md:
```diff
@@ -1,21 +1,67 @@
+
+### Citation
+
+Documentation of the associated normative and informative referneces for the codec is RECOMMENDED.
```
references

dericed replied:
fixed locally

JeromeMartinez commented on codec_specs.md:
```diff
@@ -1,21 +1,67 @@
+
+Documentation of the associated normative and informative referneces for the codec is RECOMMENDED.
+
+### Adoption Date
```
I don't think it is easy to say the adoption date because it would be actually when the RFC is accepted, and it would be redundant with the RFC release date.
I suggest to remove this part.

dericed replied:
removed Adoption Date

dericed commented:
Updated according to some suggestions from @JeromeMartinez and @mbunkus. Please re-review.

retokromer commented on codec_specs.md:
```diff
@@ -1,21 +1,61 @@
 
-*   CodecPrivate in the TrackEntry
-*   CodecState in the BlockGroup
+The intention behind this list is not to list all existing audio and video codecs, but rather to list those codecs that are currently supported in Matroska and therefore need a well defined Codec ID so that all developers supporting Matroska will use the same Codec ID. If you feel we missed support for a very important codec, please tell us on our development mailing list (cellar at ietf.org).
```
twice \`Codec ID\`

dericed replied:
updated. thx

mbunkus commented on codec_specs.md:
```diff
@@ -1,21 +1,61 @@
-The intention behind this list is `NOT` to list all existing audio and video codecs, but rather to list those codecs that are `currently supported` in Matroska and therefore need a well defined codec ID so that all developers supporting Matroska will use the same ID. If you feel we missed support for a very important codec, please tell us on our development mailing list (matroska-devel at lists.m ...
+### Codec ID
 
+Each codec supported for storage in Matroska MUST have a unique `Codec ID`. Each `Codec ID` MUST be prefixed with the string from the following table according to the associated type of the codec. All characters of a `Codec ID Prefix` MUST be capital letters (A-Z) except for the last character of a `Codec ID Prefix` which MUST be a hypen ("-").
```
Actually, the last character is not a hyphen but an underscore :)

dericed replied:
argh, right. Fixed locally.

mbunkus commented on codec_specs.md:
```diff
@@ -1,21 +1,61 @@
+
+### Initialisation
+
+Each encoding supported for storage in Matroska MUST have a defined Initialisation. The Initialisation MUST describe the storage of data necessary to initialize the decoder, which MUST be stored within the `CodecPrivate Element`. When the Initialisation is updated within a track then that updated Initialisation data MUST be written into the `CodecState Element` of the first `Cluster` to require i ...
```
I don't like the wording that an element "must be void". I'd prefer something like "must not be used" or "must be ignored" or something like that.

And there's a missing , before "then" in the last sentence.

retokromer replied:
I agree regarding «void».

dericed replied:
I agree but want the statement to affect the writer as well as the reader. I updated to:
> ...  and the `CodecPrivate Element` SHOULD NOT be written and MUST be ignored.

mbunkus commented on codec_specs.md:
```diff
@@ -1,21 +1,61 @@
+
+### Description
+
+An optional description for the encoding.
```
Maybe we should clarify that `Codec Name` and `Description` are meant solely for human consumption and have no effect on the processing of Matroska data by a demuxer.

dericed replied:
Added:
> This value is only intended for human consumption.

mbunkus commented on codec_specs.md:
```diff
@@ -1,21 +1,61 @@
+
+### Deprecation Date
+
+A timestamp, expressed in [@!RFC3339] that notes when support for the `Codec Mapping` within Matroska was deprecated. If a Codec is defined with a `Deprecation Date`, then it is RECOMMENDED that Matroska writers SHOULD NOT use the `Codec Mapping` after the `Deprecation Date`.
```
"Codec Mapping" instead of "Codec"?

dericed replied:
I used `Codec Mapping` rather than `Codec` because I didn't want saying that `A_AAC/MPEG2/LC/SBR` was deprecated could imply that `AAC` as an encoding format was deprecated. Since it is just that particular mapping of the codec to Matroska was is deprecated and not the codec itself.

mbunkus replied:
Sorry for explaining so poorly what I meant. I'm fine with `Codec Mapping`. However, you've written: `If a Codec is defined…`, and my point was that this should probably be `If a Codec Mapping is defined…` in order to be consistent.

I did not mean to advocate changing `Codec Mapping` back to `Codec`.

dericed replied:
thx for clarifying, fixed and pushed

dericed commented:
PR is adjusted to feedback and force pushed.

dericed commented:
Once again, ready for review.

robUx4 commented on codec_specs.md:
```diff
@@ -1,21 +1,61 @@
+# Codec Mappings
 
-As an additional resource to this page Haali has created a [list of codec IDs in a PDF](http://haali.su/mkv/codecs.pdf).
+A `Codec Mapping` is a set of attributes to identify, name, and contextualize the format and characteristics of encoded data that can be contained within Matroska Clusters.
```
It's actually in the `Blocks` not the `Clusters` which contain more generic stuff.

robUx4 replied:
More generally I think it should say that the codec data correspond to the Payload data found in `Block` and `SimpleBlock`,
